src/doc-printer.js

```diff
@@ -210,6 +226,77 @@ function printDocToString(doc, options) {
               break;
           }
           break;
+
+        // Fills each line with as much code as possible before moving to a new
+        // line with the same indentation.
+        //
+        // Expects doc.parts to be an array of alternating code and
+        // whitespace. The whitespace contains the linebreaks.
+        //
+        // For example:
+        //   ["I", line, "love", line, "monkeys"]
+        // or
+        //   [{ type: group, ... }, softline, { type: group, ... }]
+        //
+        // It uses this parts structure to handle three main layout cases:
+        // * The first two non-whitespace items fit on the same line without
+        //   breaking -> output both items and the whitespace "flat".
+        // * Only the first item fits on the line without breaking -> output the
+        //   first item "flat" and the whitespace with "break".
+        // * Neither item fits on the line without breaking -> output the first
+        //   item and the whitespace with "break".
+        case "fill": {
+          let rem = width - pos;
+
+          const parts = doc.parts;
+          if (parts.length === 0) {
+            break;
+          }
+
+          const first = parts[0];
+          const firstCmd = [ind, MODE_FLAT, first];
+          if (parts.length === 1) {
+            if (fits(firstCmd, cmds, width - rem, true)) {
+              cmds.push(firstCmd);
+            } else {
+              cmds.push([ind, mode, first]);
+            }
+            break;
+          }
+
+          const split = parts[1];
```

Comment:
ahh, I think I see... code like this might really clarify things for me as a reader...

```js
const split = parts[1];
assert.ok(split.type === "line" || split.type === "...");
const splitFlatCmd = [ind, MODE_FLAT, split];
const splitBreakCmd = [ind, MODE_BREAK, split];
```

I think technically that's a bit less performant in some cases because not all scenarios require the instantiation of both the Flat and Break cmd's, but the overhead should be quite low. 

I'm also not sure this will properly account for types like `break-parent`, `line-suffix`, `line-suffix-boundary`, etc – which may not appear in the JSX context but are widespread elsewhere. Perhaps that's not worth building for at this point.